java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -91,13 +91,62 @@ class DatabaseImpl(
         TODO("not implemented")
     }
 
-    override suspend fun <Data : DatabaseData> get(
+    override suspend fun get(
         storageKey: StorageKey,
-        dataType: KClass<Data>
-    ): Data? {
-        TODO("not implemented")
+        dataType: KClass<out DatabaseData>,
+        schema: Schema
+    ) = when (dataType) {
+        DatabaseData.Entity::class -> getEntity(storageKey, schema)
+        else -> TODO("Support Singletons and Collections")
     }
 
+    @VisibleForTesting
+    fun getEntity(storageKey: StorageKey, schema: Schema): DatabaseData.Entity =
+        readableDatabase.useTransaction {
+            // Fetch the entity's type by storage key.
+            val (storageKeyId, schemaTypeId) = rawQuery(
+                """
+                    SELECT storage_keys.id, entities.type_id
+                    FROM storage_keys
+                    JOIN entities
+                      ON entities.storage_key_id = storage_keys.id
+                    WHERE storage_keys.storage_key = ?
+                """.trimIndent(),
+                arrayOf(storageKey.toString())
+            ).use {
+                require(it.moveToFirst()) { "Entity at storage key $storageKey does not exist." }
+                it.getLong(0) to it.getLong(1)
+            }
+            // Fetch the entity's fields.
+            val fieldsByName = getSchemaFields(schemaTypeId)
+            val fieldsById = fieldsByName.mapKeys { it.value.fieldId }
+            // Populate the entity's field data from the database.
+            val data = mutableMapOf<FieldName, Any?>()
+            rawQuery(
+                """
+                    SELECT field_id, primitive_value_id
+                    FROM field_values
+                    WHERE entity_storage_key_id = ?
+                """.trimIndent(),
+                arrayOf(storageKeyId.toString())
+            ).forEach {
+                val fieldId = it.getLong(0)
+                val fieldValueId = it.getLong(1)
+                val field = fieldsById.getValue(fieldId)
+                // TODO: Handle non-primitive and collection field values.
+                data[field.fieldName] = getPrimitiveValue(fieldValueId, field.typeId)
```

Comment:
This actually ends up being 1 + 1 + N queries per entity, with the call to `getPrimitiveValue`. With that being the case, I think maybe a Join is better...  But not necessary in this PR.

Reply:
Added TODO